runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -1017,6 +1041,100 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 #endif /* J9VM_ENV_DATA64 */
 	}
 
+	/**
+	 * Extract the class pointer from an unforwarded object.
+	 *
+	 * This method will assert if the object has been marked as forwarded.
+	 *
+	 * @param[in] pointer to forwardedHeader the MM_ForwardedHeader instance encapsulating the object
+	 * @return pointer to the J9Class from the object encapsulated by forwardedHeader
+	 * @see MM_ForwardingHeader::isForwardedObject()
+	 */
+	MMINLINE J9Class *
+	getPreservedClass(MM_ForwardedHeader *forwardedHeader)
+	{
+		return (J9Class *)((uintptr_t)(forwardedHeader->getPreservedSlot()) & J9GC_J9OBJECT_CLAZZ_ADDRESS_MASK);
+	}
+
+	/**
+	 * Get the preserved hashcode offset of forwarded header.
+	 *
+	 * @param[in]   Pointer to forwardedHeader the MM_ForwardedHeader instance encapsulating the object
+	 */
+	MMINLINE uintptr_t
+	getPreservedHashcodeOffset(MM_ForwardedHeader *forwardedHeader)
+	{
+		J9Class *clazz = getPreservedClass(forwardedHeader);
+		return getHashcodeOffset(clazz, getPreservedArrayLayout(forwardedHeader), getPreservedIndexableSize(forwardedHeader));
+	}
+
+	/**
+	 * Extract the size (as getSizeInElements()) from an unforwarded object
+	 *
+	 * This method will assert if the object is not indexable or has been marked as forwarded.
+	 *
+	 * @param[in] forwardedHeader pointer to the MM_ForwardedHeader instance encapsulating the object
+	 * @return the size (#elements) of the array encapsulated by forwardedHeader
+	 * @see MM_ForwardingHeader::isForwardedObject()
+	 */
+	MMINLINE uint32_t
+	getPreservedIndexableSize(MM_ForwardedHeader *forwardedHeader)
+	{
+		/* Asserts if object is indeed indexable */
+		ForwardedHeaderAssert(J9GC_CLASS_IS_ARRAY(getPreservedClass(forwardedHeader)));
+
+		/* in compressed headers, the size of the object is stored in the low-order half of the uintptr_t read when we read clazz
+		 * so read it from there instead of the heap (since the heap copy would have been over-written by the forwarding
+		 * pointer if another thread copied the object underneath us). In non-compressed, this field should still be readable
+		 * out of the heap.
+		 */
+		uint32_t size = 0;
+		if (compressObjectReferences()) {
+			size = forwardedHeader->getPreservedOverlap();
+		} else {
+			size = ((J9IndexableObjectContiguousFull *)forwardedHeader->getObject())->size;
+		}
+
+		if (0 == size) {
+			/* Discontiguous */
+			if (compressObjectReferences()) {
+				size = ((J9IndexableObjectDiscontiguousCompressed *)forwardedHeader->getObject())->size;
+			} else {
+				size = ((J9IndexableObjectDiscontiguousFull *)forwardedHeader->getObject())->size;
+			}
+		}
+
+		return size;
+	}
+
+	/**
+	 * Extract the array layout from preserved info in Forwarded header
+	 * (this mimics getArrayLayout())
+	 *
+	 * @param[in] forwardedHeader pointer to the MM_ForwardedHeader instance encapsulating the object
+	 * @return the ArrayLayout for the forwarded object
+	 */
+	MMINLINE ArrayLayout
+	getPreservedArrayLayout(MM_ForwardedHeader *forwardedHeader)
+	{
```

Comment:
we should assert this is an indexable object like in the other API

Reply:
The one that we assert is actually `getPreservedIndexableSize`, should we assert here as well?

Reply:
yes